`always_treat_brackets_as_autoclosed` causes crash when deleting characters at beginning of buffer

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

With `always_treat_brackets_as_autoclosed` enabled, type in "]" and place the cursor in front of it:
![2024-07-22 14_43_05_14308](https://github.com/user-attachments/assets/9dfe4899-2b3d-497f-8434-10d15a39fe01)

Running `editor: backspace` command in this state causes a crash.

Note that the selected language must support auto-closing brackets, and that `use_autoclose` has no effect on the crash.

### Environment

Zed: v0.144.3 (Zed)
OS: Linux X11 arch unknown
Memory: 62.5 GiB
Architecture: x86_64

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.

<details><summary>Zed.log</summary><pre>
<!-- Click below this line and paste or drag-and-drop your log-->
2024-07-22T14:59:57.760543037+02:00 [INFO] ========== starting zed ==========
2024-07-22T14:59:57.769008332+02:00 [WARN] failed to load font 'SymbolsNFM'
2024-07-22T14:59:57.770514547+02:00 [WARN] failed to load font 'SymbolsNFM'
2024-07-22T14:59:57.776971773+02:00 [INFO] Compositor detection: _NET_WM_CM_S?=false, _NET_WM_CM_OWNER=false, _NET_SUPPORTING_WM_CHECK=true
2024-07-22T14:59:57.777031124+02:00 [INFO] x11: compositor present: true, gtk_frame_extents_supported: false
2024-07-22T14:59:57.780720263+02:00 [INFO] perform;
2024-07-22T14:59:57.780776962+02:00 [INFO] read_command;
2024-07-22T14:59:57.780865615+02:00 [INFO] read_command;
2024-07-22T14:59:57.781022691+02:00 [INFO] socket reader;
2024-07-22T14:59:57.781060903+02:00 [INFO] Opening main db
2024-07-22T14:59:57.781256779+02:00 [INFO] new;
2024-07-22T14:59:57.781792325+02:00 [INFO] keep_updated;
2024-07-22T14:59:57.783084812+02:00 [INFO] Using git binary path: None
2024-07-22T14:59:57.799952064+02:00 [INFO] set environment variables from shell:/usr/bin/bash, path:/usr/local/sbin:/usr/local/bin:/usr/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl:/home/XXX/scripts
2024-07-22T14:59:57.822479187+02:00 [INFO] extensions updated. loading 1, reloading 0, unloading 0
2024-07-22T14:59:57.82258493+02:00 [INFO] activate is not implemented on Linux, ignoring the call
2024-07-22T14:59:57.822631406+02:00 [INFO] Opening main db
2024-07-22T14:59:57.870407119+02:00 [INFO] perform;
2024-07-22T14:59:57.870612822+02:00 [INFO] read_command;
2024-07-22T14:59:57.870770412+02:00 [INFO] read_command;
2024-07-22T14:59:57.871077438+02:00 [INFO] socket reader;
2024-07-22T14:59:57.873348412+02:00 [INFO] Using Visual { id: 35, colormap: 0, depth: 32 }
2024-07-22T14:59:57.873563296+02:00 [INFO] Creating colormap 33554435
2024-07-22T14:59:57.877720453+02:00 [INFO] building git repository, `.git` path in the worktree: ".git"
2024-07-22T14:59:57.89773911+02:00 [INFO] Enabling Vulkan Portability
2024-07-22T14:59:57.897783194+02:00 [INFO] Enabling color space support
2024-07-22T14:59:57.936711489+02:00 [INFO] Adapter "NVIDIA GeForce RTX 2060 with Max-Q Design"
2024-07-22T14:59:57.936760446+02:00 [INFO] Ray tracing is supported
2024-07-22T14:59:58.299183885+02:00 [INFO] Using surface present mode FIFO_RELAXED
2024-07-22T14:59:58.299312725+02:00 [WARN] Unable to forbid exclusive full screen
2024-07-22T14:59:58.362078629+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-22T14:59:58.36213083+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-22T14:59:58.362170202+02:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-07-22T14:59:58.609760023+02:00 [INFO] Refreshing at 6944 micros
2024-07-22T14:59:58.609885465+02:00 [INFO] x11: no compositor present, falling back to server-side window decorations
2024-07-22T14:59:58.652337087+02:00 [WARN] GPU frame is out of date
2024-07-22T14:59:58.653054398+02:00 [WARN] Acquire failed because the surface is out of date
2024-07-22T14:59:58.65396759+02:00 [INFO] Using surface present mode FIFO_RELAXED
2024-07-22T14:59:58.65407334+02:00 [WARN] Unable to forbid exclusive full screen
2024-07-22T14:59:58.658282965+02:00 [ERROR] DBus error service error org.freedesktop.zbus.Error: The name is not activatable
2024-07-22T14:59:58.658786276+02:00 [INFO] Opening main db
2024-07-22T14:59:58.66617158+02:00 [INFO] Initializing default prettier with plugins {}
2024-07-22T14:59:58.666227476+02:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-22T14:59:58.66626254+02:00 [INFO] starting language server "json-language-server", path: "/home/XXX", id: 1
2024-07-22T14:59:58.666310175+02:00 [INFO] starting language server "package-version-server", path: "/home/XXX", id: 2
2024-07-22T14:59:58.666386574+02:00 [INFO] Initializing default prettier with plugins {}
2024-07-22T14:59:58.666410897+02:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-22T14:59:58.676751848+02:00 [INFO] fetching latest version of language server "json-language-server"
2024-07-22T14:59:58.676796117+02:00 [INFO] Node runtime install_if_needed
2024-07-22T14:59:58.691369349+02:00 [INFO] fetching latest version of language server "package-version-server"
2024-07-22T14:59:58.696325362+02:00 [INFO] Initializing default prettier with plugins {}
2024-07-22T14:59:58.696365605+02:00 [INFO] starting language servers for JSON: json-language-server, package-version-server
2024-07-22T14:59:59.312223985+02:00 [INFO] downloading language server "package-version-server"
2024-07-22T14:59:59.312353586+02:00 [INFO] starting language server. binary path: "/home/XXX/.local/share/zed/languages/package-version-server/package-version-server-v0.0.4", working directory: "/home/XXX", args: []
2024-07-22T14:59:59.654135045+02:00 [INFO] ignoring macOS specific set_edited
2024-07-22T14:59:59.846442785+02:00 [INFO] downloading language server "json-language-server"
2024-07-22T14:59:59.846922749+02:00 [INFO] Node runtime install_if_needed
2024-07-22T15:00:00.137338407+02:00 [INFO] starting language server. binary path: "/home/XXX/.local/share/zed/node/node-v18.15.0-linux-x64/bin/node", working directory: "/home/XXX", args: ["/home/XXX/.local/share/zed/languages/json-language-server/node_modules/vscode-langservers-extracted/bin/vscode-json-language-server", "--stdio"]
2024-07-22T15:00:00.808015438+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-22T15:00:00.814319644+02:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-07-22T15:00:00.818093635+02:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-07-22T15:00:00.825004805+02:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-07-22T15:00:02.322333685+02:00 [ERROR] {
  "thread": "main",
  "payload": "offset is out of range",
  "location_data": {
    "file": "crates/multi_buffer/src/multi_buffer.rs",
    "line": 4623
  },
  "backtrace": [
    "__libc_start_main"
  ],
  "app_version": "0.144.3",
  "release_channel": "Zed",
  "os_name": "Linux X11",
  "os_version": "arch unknown",
  "architecture": "x86_64",
  "panicked_on": 1721653202322,
  "installation_id": "490b21b7-a6be-4038-bfe4-32039387fc18",
  "session_id": "5bdb4c6d-b55b-4cf3-b8e2-cec2754e894c"
}
<!-- Click above this line and paste or drag-and-drop your log--></pre></details>

--- a/crates/editor/src/editor.rs
+++ b/crates/editor/src/editor.rs
@@ -4098,8 +4098,10 @@ impl Editor {
 
                         if buffer.contains_str_at(selection.start, &pair.end) {
                             let pair_start_len = pair.start.len();
-                            if buffer.contains_str_at(selection.start - pair_start_len, &pair.start)
-                            {
+                            if buffer.contains_str_at(
+                                selection.start.saturating_sub(pair_start_len),
+                                &pair.start,
+                            ) {
                                 selection.start -= pair_start_len;
                                 selection.end += pair.end.len();
 

